[WIP] Check Lock Time Verify

presstab opened the pull request:
Add checklocktimeverify to the scripting language. Enforce CLTV once 75% of the last blocks are version 5. Reject version 4 blocks once version 5 block consume 95% of the last blocks.

Mrs-X commented:
@presstab How can I test this?

Fuzzbawls commented on src/main.cpp:
```diff
@@ -3197,6 +3213,12 @@ bool ConnectBlock(const CBlock& block, CValidationState& state, CBlockIndex* pin
+    if (!fCLTVHasMajority && CBlockIndex::IsSuperMajority(5, chainActive.Tip(), Params().EnforceBlockUpgradeMajority())) {
+        fCLTVHasMajority = true;
+        LogPrintf("CHECKLOCKTIMEVERIFY achieved supermajority! Transactions that use CLTV will now be verified.\n");
+    }
```
I think this would be better suited towards the beginning of the function instead of at the end considering there is a check against `fCLTVHasMajority` further up in `ConnectBlock()`

Fuzzbawls commented on src/miner.cpp:
```diff
@@ -120,6 +120,8 @@ CBlockTemplate* CreateNewBlock(const CScript& scriptPubKeyIn, CWallet* pwallet,
     else
         pblock->nVersion = 3;
 
+    pblock->nVersion = 5;   // Supports CLTV activation
```
can we possibly clean up the preceding line's logic regarding the block version while we're at this?

Fuzzbawls commented on src/chainparams.cpp:
```diff
@@ -131,9 +131,9 @@ class CMainParams : public CChainParams
-        nToCheckBlockUpgradeMajority = 1000;
+        nEnforceBlockUpgradeMajority = 1080;
+        nRejectBlockOutdatedMajority = 1368;
+        nToCheckBlockUpgradeMajority = 1440;
```
After thinking about this for a while, I'm pretty sure that only looking at the last day's worth of blocks is much too short of a sample size, and leaves a **VERY** short window for people to update before enforcement kicks in.

Looking at the last 7 days worth of blocks (10080) probably makes more sense, with the thresholds (75%/95%) adjusted accordingly.

presstab commented:
Updated the PR to reflect suggestions. Main net activation is now based on 7 days (estimated 7 days using 7.5 days worth of blocks since blocks have seemed to come in a bit fast).

Removed databasing. Was not completely comfortable with the logic. Best plan is to do the full check as necessary. Once locked in we can replace the index iterating with a block height.
